### ISSUE TITLE
File.Delete and File.Copy

### ISSUE THREAD

**Issue by dbaldacchino** [DynamoDS/Dynamo#7431]
If this issue is with Dynamo for Revit, please post your issue on the [Dynamo for Revit Issues page](https://github.com/DynamoDS/DynamoRevit/issues).

If this issue is **not** a bug report or improvement request, please check the [Dynamo forum](https://forum.dynamobim.com/), and start a thread there to discuss your issue.

## Dynamo version

1.2.1

## Operating system

Windows 7 & Windows 10

## What did you do? 

I'm trying to understand the behavior of these nodes and perhaps make some suggestions. It appears that these work once and don't pass a result (out port is "void" and the node returns "null"). It would be valuable if these pass a result instead of null.

How can these be used in conditional situations? For example if a file exists at destination, I would like to delete it. Or if it doesn't exist, copy a new file from some source location. These nodes just take a path and I'm looking for something that triggers them to work or sit idle.

I also don't understand the order these run within a graph in relation to other nodes. Let's say I want to do the following in this order:

a) Copy an Excel template file from source to destination and give it a new name;
b) Take some data a write it to this newly named Excel file

The problem I'm running into is that the **Excel.WriteToFile** node is writing a new file before the template file has been copied and renamed, so my data goes into an empty Excel file instead. It also gives me an error because as the template file is being written with a new name, it fails since that file is already open by the Excel process which has already written data to the file and kept it open (as a side-note, it would be nice if Excel files aren't opened when read or written to, or we had control over this).

## What did you expect to see?

I'm hoping to be able to control the sequence of events as they pertain to copying, deleting and writing of/to files.

## What did you see instead?

Conflicts due to processes happening out of sequence. In a script I can control when everything happens but in Dynamo I cannot seem to have the same ability.

**Comment by dbaldacchino** [comment]
Here's an example to illustrate what I mean by conditional execution of various branches of a graph (or termination) based on certain conditions. In the example below a python script prompts the user to browse to a folder and the user can also cancel out (the pyton code does various checks and sets the OUT port accordingly). The If node only does an evaluation and sets the output, but it doesn't in itself act as a diverter to control which branch of the graph to execute. It would be very valuable to have nodes that can act as valves/diverters/switches to direct graph execution to a specific branch or halt it altogether. I don't think this is possible, correct? Any tips/feedback would be greatly appreciated, thanks.

![conditional execution](https://cloud.githubusercontent.com/assets/4656191/21555555/b83350a2-cdde-11e6-8ec8-db7a6aab93ef.png)

**Comment by mjkkirschner** [comment]
getting control flow into a data flow language is pretty difficult - Dynamo's UI could support control flow but currently does not. You're always (almost) passing data around.

To order execution now you can use function composition - this is a headache to construct in the graph, but it works. 

There are a few nodes that don't exactly pass data - like the `scope if` node - this node will pass code around, and will do what you want, only one brach of logic will get executed based on the test. It has a lot of gotchas though - like the branches cannot share data.

The easiest way to build control flow and real conditional workflows right now is to use code blocks with [imperative] blocks, where you can use if statements - or python, etc.... Getting back into the graph after that point is again tough :(.

To solve your excel issue I would look at function.compose to order the nodes to run in an explicit order - you could also do this by making custom nodes that take some dummy input and output and passing it around in order. Or -  call these functions from design script in an imperative block in order.

**Comment by dbaldacchino** [comment]
Thanks Michael, unfortunately there's virtually no info. on the use of the ScopeIf node (except some posts on GitHub, where most point out the node was intended for recursion inside custom nodes and shouldn't be used in graphs). I think I understand the concept you outline using Imperative blocks after searching around a bit. Would be nice to see some blog posts and more documentation on these techniques.

The problem I'm having is one of "cleanliness" really. The graph executes either way, but when coding/scripting, you can handle whether something executes or not. In the case of Dynamo, I'd like more control over that to avoid errors (yellow nodes). This is especially important when using Dynamo Player, where it looks sloppy when execution finishes with errors. So for example if I use a Python node to ask the user to select a file to be copied or deleted and they cancel out, I want to stop trying to copy/delete because the user input does not contain the proper path required by the downstream nodes. This is just a simple example that illustrates what I'm after. Another example is when writing data to an Excel file: if I have no data to write, the branch that creates and opens the Excel file still executes, which again is a bit sloppy.

As to the function.compose, I have only used this sparingly in the past in addition to list.map after seeing some examples that made sense to me, although I have to admit that I don't have a solid understanding of how it works. Any tips/examples would be greatly appreciated to get me started in the right direction. Thanks again!

**Comment by dbaldacchino** [comment]
Hi again Michael, after playing around a bit with the ScopeIf node, it does appear it might do the job. However as you pointed out, you get a lot of errors and failures if you have shared data between branches, so I'm not sure it'll get me what I want; perhaps just a little closer (might have to live with errors but at least this will possibly prevent Excel from opening a workbook in the second example cited above when a user cancels).

**Comment by andydandy74** [comment]
@dbaldacchino FWIW, I usually juts pass ```null``` if I want to keep a branch from dpoing anything meaningful. It's a bit on the quick-and-dirty side but it generally works quite well.

**Comment by dbaldacchino** [comment]
Thanks @andydandy74, unfortunately there are some nodes in the definition that throw an error regardless. Having said that though, using a ScopeIf node in a custom node (wrapped the Excel.Write node as shown below) helps to clean things up a bit and avoids having Excel open up and write non-relevant data to a file that needs to be discarded anyway :)

![scopeif_customnode](https://cloud.githubusercontent.com/assets/4656191/21623550/84a70812-d1c8-11e6-9c5a-b8b754ec8d39.png)

I am finding that using ScopeIf in a definition is causing unexpected behavior. For example I would expect that the test branch would evaluate first, but it's throwing an error that there's an undefined variable (a python node connected to both the test and true branches isn't evaluating, hence the ScopeIf node throws an error). If I substitute it with an If node instead, it works. So the question in this case is: should the ScopeIf node be used in a definition or should it stricly be confined to Custom Nodes only? Do you need example files that demonstrate this behavior? Thanks.

**Comment by dbaldacchino** [comment]
Well I finally managed to get a definition that runs without errors even if the user cancels out of the TaskFialog or FolderBrowserDialog, but boy this was way more difficult than it should be (hours upon hours wasted on trying to get this to work and only because I'm stubborn!). I used a couple of additional nodes (thanks @andydandy74 for Clockwork) to replace an empty list with something that List.Deconstruct can work with. I was not able to achieve this with any out of the box nodes either. Thanks again.

![solution](https://cloud.githubusercontent.com/assets/4656191/21624774/f18bfe10-d1cd-11e6-8bff-10b956f41fdb.png)
